R/pmcmc_parallel.R

```diff
@@ -0,0 +1,295 @@
+      private$sessions <- vector("list", control$n_workers)
+      private$status <- vector("list", control$n_chains)
+      private$results <- vector("list", control$n_chains)
+
+      ## First stage starts the process, but this is async...
+      for (i in seq_len(control$n_workers)) {
+        private$remotes[[i]] <- pmcmc_remote$new(
+          pars, initial, inputs, control, seed)
+        private$sessions[[i]] <- private$remotes[[i]]$session
+      }
+      ## ...so once the sessions start coming up we start them working
+      for (i in seq_len(control$n_workers)) {
+        private$remotes[[i]]$wait_session_ready()
+        private$status[[i]] <- private$remotes[[i]]$init(i)
+      }
+    },
+
+    step = function() {
+      res <- processx::poll(private$sessions, 1000)
+      i <- vcapply(res, "[[", "process") == "ready"
+      if (any(i)) {
+        dat <- lapply(private$remotes[i], function(x) x$read())
+        index <- vnapply(dat, "[[", "index")
+        result <- lapply(dat, "[[", "result")
+        private$status[index] <- result
+        finished <- vlapply(result, function(x) x$finished)
+        for (r in private$remotes[i][!finished]) {
+          private$thread_pool$remove(r)
+          r$continue()
+        }
+        if (any(finished)) {
+          ## It might be preferable to close out sessions *first* as
+          ## if we close out 2 sessions here simultaneously we
+          ## underallocate cores for one time-chunk. That does
+          ## complicate the book-keeping though.
+          remaining <- which(lengths(private$status) == 0)
+          for (r in private$remotes[i][finished]) {
+            res <- r$finish()
+            private$results[res$index] <- list(res$data)
+            if (length(remaining) == 0L) {
+              r$session$close()
+              private$thread_pool$add(r)
+            } else {
+              private$thread_pool$remove(r)
+              j <- remaining[[1]]
+              remaining <- remaining[-1]
+              private$status[[j]] <- r$init(j)
+            }
+          }
+        }
+      }
+      private$progress(private$status)
+    },
+
+    run = function() {
+      while (!self$step()) {
+      }
+    },
+
+    finish = function() {
+      pmcmc_combine(samples = private$results)
+    }
+  ))
+
+
+pmcmc_remote <- R6::R6Class(
+  "pmcmc_remote",
+  private = list(
+    pars = NULL,
+    initial = NULL,
+    inputs = NULL,
+    control = NULL,
+    seed = NULL,
+    step = NULL
+  ),
+
+  public = list(
+    session = NULL,
+    index = NULL,
+    n_threads = NULL,
+
+    initialize = function(pars, initial, inputs, control, seed) {
+      self$session <- callr::r_session$new(wait = FALSE)
+
+      private$pars <- pars
+      private$initial <- initial
+      private$inputs <- inputs
+      private$control <- control
+      private$seed <- seed
+
+      lockBinding("session", self)
+    },
+
+    wait_session_ready = function(timeout = 3000) {
```

Comment:
This is the timeout before failure; if the session does not come alive in 3s we assume that it will not come up and and error will be thrown. 3s was the default used in the callr's higher-level functions